README.md
# <img alt='' src='https://gitea.com/repo-avatars/550-80a3a8c2ab0e2c2d69f296b7f8582485' height="40"/> *T E A*

[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT) [![Release](https://raster.shields.io/badge/dynamic/json.svg?label=release&url=https://gitea.com/api/v1/repos/gitea/tea/releases&query=$[0].tag_name)](https://gitea.com/gitea/tea/releases) [![Build Status](https://drone.gitea.com/api/badges/gitea/tea/status.svg)](https://drone.gitea.com/gitea/tea) [![Join the chat at https://img.shields.io/discord/322538954119184384.svg](https://img.shields.io/discord/322538954119184384.svg)](https://discord.gg/Gitea) [![Go Report Card](https://goreportcard.com/badge/code.gitea.io/tea)](https://goreportcard.com/report/code.gitea.io/tea) [![GoDoc](https://godoc.org/code.gitea.io/tea?status.svg)](https://godoc.org/code.gitea.io/tea)

### The official CLI for Gitea

![demo gif](./demo.gif)

```
   tea - command line tool to interact with Gitea
   version 0.7.0-preview

 USAGE
   tea command [subcommand] [command options] [arguments...]

 DESCRIPTION
   tea is a productivity helper for Gitea.  It can be used to manage most entities on one
   or multiple Gitea instances and provides local helpers like 'tea pull checkout'.
   tea makes use of context provided by the repository in $PWD if available, but is still
   usable independently of $PWD. Configuration is persisted in $XDG_CONFIG_HOME/tea.

 COMMANDS
   help, h  Shows a list of commands or help for one command
   ENTITIES:
     issues, issue, i                  List, create and update issues
     pulls, pull, pr                   Manage and checkout pull requests
     labels, label                     Manage issue labels
     milestones, milestone, ms         List and create milestones
     releases, release, r              Manage releases
     times, time, t                    Operate on tracked times of a repository's issues & pulls
     organizations, organization, org  List, create, delete organizations
     repos, repo                       Show repository details
   HELPERS:
     open, o                         Open something of the repository in web browser
     notifications, notification, n  Show notifications
   SETUP:
     logins, login                  Log in to a Gitea server
     logout                         Log out from a Gitea server
     shellcompletion, autocomplete  Install shell completion for tea

 OPTIONS
   --help, -h     show help (default: false)
   --version, -v  print the version (default: false)

 EXAMPLES
   tea login add                       # add a login once to get started

   tea pulls                           # list open pulls for the repo in $PWD
   tea pulls --repo $HOME/foo          # list open pulls for the repo in $HOME/foo
   tea pulls --remote upstream         # list open pulls for the repo pointed at by
                                       # your local "upstream" git remote
   # list open pulls for any gitea repo at the given login instance
   tea pulls --repo gitea/tea --login gitea.com

   tea milestone issues 0.7.0          # view open issues for milestone '0.7.0'
   tea issue 189                       # view contents of issue 189
   tea open 189                        # open web ui for issue 189
   tea open milestones                 # open web ui for milestones

   # send gitea desktop notifications every 5 minutes (bash + libnotify)
   while :; do tea notifications --mine -o simple | xargs -i notify-send {}; sleep 300; done

 ABOUT
   Written & maintained by The Gitea Authors.
   If you find a bug or want to contribute, we'll welcome you at https://gitea.com/gitea/tea.
   More info about Gitea itself on https://gitea.io.
```

- [Compare features with other git forge CLIs](./FEATURE-COMPARISON.md)
- tea uses [code.gitea.io/sdk](https://code.gitea.io/sdk) and interacts with the Gitea API.

## Installation

There are different ways to get `tea`:

1. Install via your system package manager:
    - macOS via `brew` (gitea-maintained):
      ```sh
      brew tap gitea/tap https://gitea.com/gitea/homebrew-gitea
      brew install tea
      ```
    - arch linux ([gitea-tea](https://aur.archlinux.org/packages/gitea-tea), thirdparty)
    - alpine linux ([tea](https://pkgs.alpinelinux.org/packages?name=tea&branch=edge), thirdparty)

2. Use the prebuilt binaries from [dl.gitea.io](https://dl.gitea.io/tea/)

3. Install from source (go 1.13 or newer is required):
    ```sh
    go get code.gitea.io/tea
    go install code.gitea.io/tea
    ```

4. Docker (thirdparty): [tgerczei/tea](https://hub.docker.com/r/tgerczei/tea)

## Compilation

Make sure you have installed a current go version.
To compile the sources yourself run the following:

```sh
git clone https://gitea.com/gitea/tea.git
cd tea
make STATIC=true
```

## Contributing

Fork -> Patch -> Push -> Pull Request

- `make test` run testsuite
- `make vet`  run checks (check the order of imports; preventing failure on CI pipeline beforehand)
- `make vendor` when adding new dependencies
- ... (for other development tasks, check the `Makefile`)

**Please** read the [CONTRIBUTING](CONTRIBUTING.md) documentation, it will tell you about internal structures and concepts.

## License

This project is under the MIT License. See the [LICENSE](LICENSE) file for the
full license text.


CHANGELOG.md
# Changelog

## [v0.8.0](https://gitea.com/gitea/tea/releases/tag/v0.8.0) - 2021-09-22

* BREAKING
  * `tea notifications --all` has moved to `tea notifications --mine` (#389)
  * `tea notifications` now only works with the context of a remote repo. (#389)
    To run this outside of a local git dir, run either tea n `--mine` or `tea n --repo <my/repo>`
* FEATURES
  * Add `tea pr merge` (#348)
* BUGFIXES
  * Don't skip reading the local repo when `--repo` specifies a repo slug (#398)
  * Fix adding login without token on private instances (#392)
  * Correctly match login by ssh host with port (#391)
  * Fix printing issue deadline (#388)
  * Return useful error on wrong sshkey path (#374)
  * Fix parsing of `--description` for issue/pr create (#371)
  * Add missing flags (#369)
  * Check negative limit command parameter (#358) (#359)
  * Add missing flags to org & labels subcommands (#357)
* ENHANCEMENTS
  * Don't require a body for comment PR reviews (#399)
  * Accept more main branch names for login detection (#396)
  * Make local repo optional for `tea pr create`(#393)
  * Notifications Add State Field (#384)
  * Improve error messages (#370)
  * Add tab completion for fish shell (#364)
  * Text editor selection: follow unix defacto standards (#356)
* MISC
  * Update Dependencies (#390)

## [v0.7.1](https://gitea.com/gitea/tea/releases/tag/v0.7.1) - 2021-08-27

* BUILD
  * Enable release builds for darwin/arm64 (#360)

## [v0.7.0](https://gitea.com/gitea/tea/releases/tag/v0.7.0) - 2021-03-12

* BREAKING
  * `tea issue create`: move `-b` flag to `-d` (#331)
  * Drop `tea notif` shorthand in favor of `tea n` (#307)
* FEATURES
  * Add commands for reviews (#315)
  * Add `tea comment` and show comments of issues/pulls (#313)
  * Add interactive mode for `tea milestone create` (#310)
  * Add command to install shell completion (#309)
  * Implement PR closing and reopening (#304)
  * Add interactive mode for `tea issue create` (#302)
* BUGFIXES
  * Introduce workaround for missing pull head sha (#340)
  * Don't exit if we can't find a local repo with a remote matching to a login (#336)
  * Don't push before creating a pull (#334)
  * InitCommand() robustness (#327)
  * `tea comment`: handle piped stdin (#322)
* ENHANCEMENTS
  * Allow checking out PRs with deleted head branch (#341)
  * Markdown renderer: detect terminal width, resolve relative URLs (#332)
  * Add more issue / pr creation parameters (#331)
  * Improve `tea time` (#319)
  * `tea pr checkout`: dont create local branches (#314)
  * Add `tea issues --fields`, allow printing labels (#312)
  * Add more command shorthands (#307)
  * Show PR CI status (#306)
  * Make PR workflow helpers more robust (#300)

## [v0.6.0](https://gitea.com/gitea/tea/releases/tag/v0.6.0) - 2020-12-11

* BREAKING
  * Add `tea repos search`, improve repo listing (#215)
  * Add Detail View for Login (#212)
* FEATURES
  * Add interactive mode for `tea pr create` (#279)
  * Add organization delete command (#270)
  * Add organization list command (#264)
* BUGFIXES
  * Forces needed arguments to `tea ms issues` (#297)
  * Subcommands work outside of git repos (#285)
  * Fix repo flag ignores local repo for login detection (#285)
  * Improve ssh handling (#277)
  * Issue create return web url (#257)
  * Support prerelease gitea instances (#252)
  * Fix `tea pr create` within same repo (#248)
  * Handle login name case-insensitive on all comands (#227)
* ENHANCEMENTS
  * Add `tea login delete` (#296)
  * Release delete: add --delete-tag & --confirm (#286)
  * Sorted milestones list (#281)
  * Pull clean & checkout use token for http(s) auth (#275)
  * Show more infos in pull detail view (#271)
  * Specify fields to print on `tea repos list` (#223)
  * Print times in local timezone (#217)
  * Issue create/edit print details (#214)
  * Improve `tea logout` (#213)
  * Added a shorthand for notifications (#209)
  * Common subcommand naming scheme (#208)
  * `tea pr checkout`: fetch via ssh if available (#192)
  * Major refactor of codebase
* BUILD
  * Use gox to cross-compile (#274)
* DOCS
  * Update Docs to new code structure (#247)

## [v0.5.0](https://gitea.com/gitea/tea/releases/tag/v0.5.0) - 2020-09-27

* BREAKING
  * Add Login Manage Functions (#182)
* FEATURES
  * Add Release Subcomands (#195)
  * Render Markdown and colorize labels table (#181)
  * Add BasicAuth & Interactive for Login (#174)
  * Add milestones subcomands (#149)
* BUGFIXES
  * Fix Pulls Create (#202)
  * Pulls create: detect head branch repo owner (#193)
  * Fix Labels Delete (#180)
* ENHANCEMENTS
  * Add Pagination Options for List Subcomands (#204)
  * Issues/Pulls: Details show State (#196)
  * Make issues & pulls subcommands consistent (#188)
  * Update SDK to v0.13.0 (#179)
  * More Options To Specify Repo (#178)
  * Add Repo Create subcomand & enhancements (#173)
  * Times: format duration as seconds for machine-readable outputs (#168)
  * Add user message to login list view (#166)

## [v0.4.1](https://gitea.com/gitea/tea/releases/tag/v0.4.1) - 2020-09-13

* BUGFIXES
  * Notification don't relay on a repo (#159)

## [v0.4.0](https://gitea.com/gitea/tea/pulls?q=&type=all&state=closed&milestone=1264) - 2020-07-18

* FEATURES
  * Add notifications subcomand (#148)
  * Add subcomand 'pulls create' (#144)
* BUGFIXES
  * Fix Login Detection By Repo Param (#151)
  * Fix Login List Output (#150)
  * Fix --ssh-key Option (#135)
* ENHANCEMENTS
  * Subcomand Login Show List By Default (#152)
* BUILD
  * Migrate src-d/go-git to go-git/go-git (#128)
  * Migrate gitea-sdk to v0.12.0 (#133)
  * Migrate yaml lib (#130)
  * Add gitea-vet (#121)

## [v0.3.1](https://gitea.com/gitea/tea/pulls?q=&type=all&state=closed&milestone=1265) - 2020-06-15

* BUGFIXES
  * --ssh-key should be string not bool (#135) (#137)
  * modules/git: fix dropped error (#127)
  * Issues details: add missing newline (#126)

## [v0.3.0](https://gitea.com/gitea/tea/pulls?q=&type=all&state=closed&milestone=1227) - 2020-04-22


main.go
// Copyright 2020 The Gitea Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

// Tea is command line tool for Gitea.
package main // import "code.gitea.io/tea"

import (
	"fmt"
	"os"
	"strings"

	"code.gitea.io/tea/cmd"

	"github.com/urfave/cli/v2"
)

// Version holds the current tea version
var Version = "development"

// Tags holds the build tags used
var Tags = ""

func main() {
	// make parsing tea --version easier, by printing /just/ the version string
	cli.VersionPrinter = func(c *cli.Context) { fmt.Fprintln(c.App.Writer, c.App.Version) }

	app := cli.NewApp()
	app.Name = "tea"
	app.Usage = "command line tool to interact with Gitea"
	app.Description = appDescription
	app.CustomAppHelpTemplate = helpTemplate
	app.Version = Version + formatBuiltWith(Tags)
	app.Commands = []*cli.Command{
		&cmd.CmdLogin,
		&cmd.CmdLogout,
		&cmd.CmdAutocomplete,

		&cmd.CmdIssues,
		&cmd.CmdPulls,
		&cmd.CmdLabels,
		&cmd.CmdMilestones,
		&cmd.CmdReleases,
		&cmd.CmdTrackedTimes,
		&cmd.CmdOrgs,
		&cmd.CmdRepos,
		&cmd.CmdAddComment,

		&cmd.CmdOpen,
		&cmd.CmdNotifications,
	}
	app.EnableBashCompletion = true
	err := app.Run(os.Args)
	if err != nil {
		// app.Run already exits for errors implementing ErrorCoder,
		// so we only handle generic errors with code 1 here.
		fmt.Fprintf(app.ErrWriter, "Error: %v\n", err)
		os.Exit(1)
	}
}

func formatBuiltWith(Tags string) string {
	if len(Tags) == 0 {
		return ""
	}

	return " built with: " + strings.Replace(Tags, " ", ", ", -1)
}

var appDescription = `tea is a productivity helper for Gitea.  It can be used to manage most entities on one
or multiple Gitea instances and provides local helpers like 'tea pull checkout'.
tea makes use of context provided by the repository in $PWD if available, but is still
usable independently of $PWD. Configuration is persisted in $XDG_CONFIG_HOME/tea.
`

var helpTemplate = bold(`
   {{.Name}}{{if .Usage}} - {{.Usage}}{{end}}`) + `
   {{if .Version}}{{if not .HideVersion}}version {{.Version}}{{end}}{{end}}

 USAGE
   {{if .UsageText}}{{.UsageText}}{{else}}{{.HelpName}}{{if .Commands}} command [subcommand] [command options]{{end}} {{if .ArgsUsage}}{{.ArgsUsage}}{{else}}[arguments...]{{end}}{{end}}{{if .Description}}

 DESCRIPTION
   {{.Description | nindent 3 | trim}}{{end}}{{if .VisibleCommands}}

 COMMANDS{{range .VisibleCategories}}{{if .Name}}
   {{.Name}}:{{range .VisibleCommands}}
     {{join .Names ", "}}{{"\t"}}{{.Usage}}{{end}}{{else}}{{range .VisibleCommands}}
   {{join .Names ", "}}{{"\t"}}{{.Usage}}{{end}}{{end}}{{end}}{{end}}{{if .VisibleFlags}}

 OPTIONS
   {{range $index, $option := .VisibleFlags}}{{if $index}}
   {{end}}{{$option}}{{end}}{{end}}

 EXAMPLES
   tea login add                       # add a login once to get started

   tea pulls                           # list open pulls for the repo in $PWD
   tea pulls --repo $HOME/foo          # list open pulls for the repo in $HOME/foo
   tea pulls --remote upstream         # list open pulls for the repo pointed at by
                                       # your local "upstream" git remote
   # list open pulls for any gitea repo at the given login instance
   tea pulls --repo gitea/tea --login gitea.com

   tea milestone issues 0.7.0          # view open issues for milestone '0.7.0'
   tea issue 189                       # view contents of issue 189
   tea open 189                        # open web ui for issue 189
   tea open milestones                 # open web ui for milestones

   # send gitea desktop notifications every 5 minutes (bash + libnotify)
   while :; do tea notifications --mine -o simple | xargs -i notify-send {}; sleep 300; done

 ABOUT
   Written & maintained by The Gitea Authors.
   If you find a bug or want to contribute, we'll welcome you at https://gitea.com/gitea/tea.
   More info about Gitea itself on https://gitea.io.
`

func bold(t string) string {
	return fmt.Sprintf("\033[1m%s\033[0m", t)
}
